actionpack/lib/action_controller/metal/strong_parameters.rb

```diff
@@ -868,13 +868,28 @@ def merge(other_hash)
       )
     end
 
+    # Returns a new <tt>ActionController::Parameters</tt> with all keys from
+    # +other_hash+ merged recursively into current hash.
+    def deep_merge(other_hash)
+      new_instance_with_inherited_permitted_status(
+        @parameters.deep_merge(other_hash.to_h)
+      )
+    end
```

Comment:
Because of the way [`deep_merge` is implemented](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/hash/deep_merge.rb#L25-L26), this can yield surprising results with nested permitted params:

```ruby
params = ActionController::Parameters.new({
  person: {
    contact: {
      email: "foo@example.com",
      blah: "blah",
    },
  }
})
#=> #<ActionController::Parameters {"person"=>{"contact"=>{"email"=>"foo@example.com", "blah"=>"blah"}}} permitted: false>

permitted = params.permit({ person: { contact: [:email] } })
# => #<ActionController::Parameters {"person"=>#<ActionController::Parameters {"contact"=>#<ActionController::Parameters {"email"=>"foo@example.com"} permitted: true>} permitted: true>} permitted: true>

params.deep_merge({ person: { foo: { bar: "baz" } } })
# => #<ActionController::Parameters {"person"=>{"contact"=>{"email"=>"foo@example.com", "blah"=>"blah"}, "foo"=>{"bar"=>"baz"}}} permitted: false>

permitted.deep_merge({ person: { foo: { bar: "baz" } } })
# => #<ActionController::Parameters {"person"=>{"foo"=>{"bar"=>"baz"}}} permitted: true>
```

We would need to override the default `deep_merge` conflict-resolution behavior with something like:

```suggestion
    def deep_merge(other_hash)
      new_instance_with_inherited_permitted_status(
        @parameters.deep_merge(other_hash.to_h) do |key, this_val, other_val|
          if (this_val.is_a?(Hash) || this_val.is_a?(Parameters)) 
              && (other_val.is_a?(Hash) || other_val.is_a?(Parameters))
            this_val.deep_merge(other_val.to_h)
          else
            other_val
          end
        end
      )
    end
```

Which is... not pretty. :disappointed:  (And likewise for `deep_merge!`.)

Also, we would definitely need more tests for cases like these.

Reply:
Thank you for sharing this extra context! As an alternative to making `ActionController::Parameters#deep_merge` more robust, would it be worthwhile to change the [Hash#deep_merge!](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/hash/deep_merge.rb#L25-L26) implementation to utilize duck typing instead of inheritance?

```diff
  def deep_merge!(other_hash, &block)
    merge!(other_hash) do |key, this_val, other_val|
-     if this_val.is_a?(Hash) && other_val.is_a?(Hash)
+     if this_val.responds_to?(:deep_merge) && other_val.responds_to?(:deep_merge)
        this_val.deep_merge(other_val, &block)
      elsif block_given?
        block.call(key, this_val, other_val)
      else
        other_val
      end
    end
  end
```

That'll make this changeset more terse at the expense of increasing the potential for farther-reaching side-effects.

Reply:
> at the expense of increasing the potential for farther-reaching side-effects.

Yes, `respond_to?(:deep_merge)` might be too permissive.  For example, with that change:

```ruby
# simplified Array#deep_merge:
Array.define_method(:deep_merge){ |other| other + self[other.size..].to_a }

{ foo: [] }.deep_merge(foo: {})
# => undefined method `+' for {}:Hash (NoMethodError)

{ foo: {} }.deep_merge(foo: [])
# => no implicit conversion of Array into Hash (TypeError)
```

But we may be able to abstract the deep merge algorithm so that it can be used in other classes.  To that end, I've submitted #45411.